Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.mm

```diff
@@ -0,0 +1,601 @@
+{
+    ASSERT(isMainThread());
+    ASSERT(!m_audioCompressor);
+    ASSERT(!m_videoEncoder);
+}
+
+void MediaRecorderPrivateWriterWebM::close()
+{
+    m_audioCompressor = nullptr;
+    m_videoEncoder = nullptr;
+}
+
+void MediaRecorderPrivateWriterWebM::compressedAudioOutputBufferCallback(void* mediaRecorderPrivateWriter, CMBufferQueueTriggerToken)
+{
+    LOG(MediaStream, "compressedAudioOutputBufferCallback called");
+    callOnMainThread([weakWriter = ThreadSafeWeakPtr<MediaRecorderPrivateWriterWebM> { static_cast<MediaRecorderPrivateWriterWebM*>(mediaRecorderPrivateWriter) }] {
+        if (auto strongWriter = weakWriter.get()) {
+            strongWriter->enqueueCompressedAudioSampleBuffers();
+            strongWriter->partiallyFlushEncodedQueues();
+        }
+    });
+}
+
+bool MediaRecorderPrivateWriterWebM::initialize(const MediaRecorderPrivateOptions& options)
+{
+    m_videoCodec = 'vp08';
+    m_audioCodec = kAudioFormatOpus;
+    ContentType mimeType(options.mimeType);
+
+    for (auto& codec : mimeType.codecs()) {
+        if (codec.startsWith("vp09"_s) || equal(codec, "vp9"_s) || equal(codec, "vp9.0"_s))
+            m_videoCodec = kCMVideoCodecType_VP9;
+        else if (codec.startsWith("vp08"_s) || equal(codec, "vp8"_s) || equal(codec, "vp8.0"_s))
+            m_videoCodec = 'vp08';
+        else if (codec == "opus"_s)
+            m_audioCodec = kAudioFormatOpus;
+        else
+            return false; // unsupported codec.
+    }
+
+    if (m_hasAudio) {
+        m_audioCompressor = AudioSampleBufferCompressor::create(m_audioCodec, compressedAudioOutputBufferCallback, this);
+        if (!m_audioCompressor)
+            return false;
+        if (options.audioBitsPerSecond)
+            RefPtr { m_audioCompressor }->setBitsPerSecond(*options.audioBitsPerSecond);
+    }
+
+    if (options.videoBitsPerSecond)
+        m_videoBitsPerSecond = *options.videoBitsPerSecond;
+
+    m_mimeType = makeString(m_hasVideo ? "video/webm"_s : "audio/webm"_s, "; codecs=\""_s, m_hasVideo ? codecStringForMediaVideoCodecId(m_videoCodec) : ""_s, m_hasVideo && m_hasAudio ? ","_s : ""_s, m_hasAudio ? codecStringForMediaVideoCodecId(m_audioCodec) : ""_s, "\""_s);
+    return true;
+}
+
+void MediaRecorderPrivateWriterWebM::enqueueCompressedAudioSampleBuffers()
+{
+    ASSERT(m_hasAudio);
+
+    if (!m_audioCompressor || m_audioCompressor->isEmpty()) {
+        LOG(MediaStream, "enqueueCompressedAudioSampleBuffers: empty queue");
+        return;
+    }
+
+    if (!m_audioFormatDescription) {
+        m_audioFormatDescription = PAL::CMSampleBufferGetFormatDescription(RefPtr { m_audioCompressor }->getOutputSampleBuffer());
+
+        const AudioStreamBasicDescription* const asbd = PAL::CMAudioFormatDescriptionGetStreamBasicDescription(m_audioFormatDescription.get());
+        if (!asbd)
+            return;
+
+        m_audioTrackIndex = m_delegate->addAudioTrack(*asbd, mkvCodeIcForMediaVideoCodecId(m_audioCodec));
```

Comment:
I would tend to pass m_audioCodec here and do the conversion at libwebm call site.
Ditto for video.